examples/sp2_gcl/node_main.py

```diff
@@ -0,0 +1,140 @@
+        c = list(c)
+        components.append(c)
+        lens.append(len(c))
+
+    return lens, components
+def compute_laplacian(data):
+
+    edge_index = data.edge_index
+    num_nodes = data.num_nodes
+    row, col = edge_index
+    data_adj = csr_matrix((np.ones(len(row)), (row, col)), shape=(num_nodes, num_nodes))
+    degree = np.array(data_adj.sum(axis=1)).flatten()
+    deg_inv_sqrt = 1.0 / np.sqrt(degree)
+    deg_inv_sqrt[np.isinf(deg_inv_sqrt)] = 0
+    I = csr_matrix(np.eye(num_nodes))
+    D_inv_sqrt = csr_matrix((deg_inv_sqrt, (np.arange(num_nodes), np.arange(num_nodes))))
+    L = I - D_inv_sqrt.dot(data_adj).dot(D_inv_sqrt)
+    e, u = scipy.sparse.linalg.eigsh(L, k=100, which='SM', tol=1e-3)
+    adj = to_scipy_sparse_matrix(data.edge_index)
+    lens, components = connected_components(adj)
+    data.e = tlx.convert_to_tensor(e, dtype=tlx.float32)
+    data.u = tlx.convert_to_tensor(u, dtype=tlx.float32)
+
+    return data, lens, components
+
+
+class ContrastiveLoss(WithLoss):
+    def __init__(self, model, temp=1.0):
+        super(ContrastiveLoss, self).__init__(backbone=model, loss_fn=None)
+        self.temp = temp
+
+    def forward(self, data, label):
+        h_node_spa, h_node_spe = self.backbone_network(data['x'], data['edge_index'], data['e'], data['u'])
+        h1 = tlx.l2_normalize(h_node_spa, axis=-1, eps=1e-12)
+        h2 = tlx.l2_normalize(h_node_spe, axis=-1, eps=1e-12)
+        logits = tlx.matmul(h1, h2.transpose(-2, -1)) / self.temp
+        labels = tlx.arange(start=0, limit=h1.shape[0], delta=1, dtype=tlx.int64)
+        loss = 0.5 * tlx.losses.softmax_cross_entropy_with_logits(logits, labels) + 0.5 * tlx.losses.softmax_cross_entropy_with_logits(logits.transpose(-2, -1), labels)
+        return loss
+def main(args):
+    global edge, e, u, test_idx
+    print(args.dataset)
+    if args.dataset in ['pubmed-3', 'flickr', 'arxiv', 'wiki', 'facebook']:
+        dataset = FacebookPagePage(root='data/facebook')
+        data = dataset[0]
+        data, lens, components = compute_laplacian(data)
+        x = tlx.convert_to_tensor(data.x, dtype=tlx.float32)
+        edge = tlx.convert_to_tensor(data.edge_index, dtype=tlx.int64)
+        e = tlx.convert_to_tensor(data.e[:args.spe_dim], dtype=tlx.float32)
+        u = tlx.convert_to_tensor(data.u[:, :args.spe_dim], dtype=tlx.float32)
+        y = tlx.convert_to_tensor(data.y)
+        print(y.min().item(), y.max().item())
+        if 'train_mask' in data.keys:
+            if len(data.train_mask.size()) > 1:
+                train_idx = tlx.where(data.train_mask[:, args.seed])[0]
+                val_idx = tlx.where(data.val_mask[:, args.seed])[0]
+                test_idx = tlx.where(data.test_mask)[0]
+            else:
+                train_idx = tlx.where(data.train_mask)[0]
+                val_idx = tlx.where(data.val_mask)[0]
+                test_idx = tlx.where(data.test_mask)[0]
+        else:
+            train_idx, val_idx, test_idx = split(y)
```

Comment:
I do not think this is useful. Usually, the train, valid, test split should be done in the dataset. You may directly use `data.train_mask` .etc to get the idx instead of add a new function in the `util`.